src/main/java/hudson/plugins/ec2/util/EC2KeyPairManager.java

```diff
@@ -0,0 +1,64 @@
+package hudson.plugins.ec2.util;
+
+import com.amazonaws.AmazonClientException;
+import com.amazonaws.services.ec2.AmazonEC2;
+import com.amazonaws.services.ec2.model.CreateKeyPairRequest;
+import com.amazonaws.services.ec2.model.CreateKeyPairResult;
+import com.amazonaws.services.ec2.model.DeleteKeyPairRequest;
+import com.amazonaws.services.ec2.model.KeyPair;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.Extension;
+import hudson.plugins.ec2.EC2AbstractSlave;
+import hudson.plugins.ec2.EC2Cloud;
+import hudson.util.Secret;
+
+import java.util.HashMap;
+import java.util.UUID;
+import java.util.logging.Logger;
+
+@Extension
+public class EC2KeyPairManager {
+    private HashMap<String, Secret> vault;
```

Comment:
> This is not apparently persisted, so it seems there is no way this could work across controller restarts? Nor would it work in a CloudBees CI HA controller.

I am aware, i need some advice on the best way to achieve that (although i don't see why it would not be possible)

Reply:
> the best way to achieve that

By keeping the `Secret` inside the `Slave` / `ComputerLauncher`, as in #993, and not in some global map.

Reply:
the convience methods in `ec2Cloud` call `save()` after creating/deleting a keypair, AFAICT this does persist across restarts

Reply:
> the [convenience] methods in `ec2Cloud` call `save()`

On `EC2Cloud.DescriptorImpl`? Which has no instance fields and is unrelated to the `EC2KeyPairManager` singleton?

Oh I see, https://github.com/jenkinsci/ec2-plugin/blob/69ccf14155a7b06b1ad0675baaff2ca00c3f3f0b/src/main/java/hudson/plugins/ec2/EC2Cloud.java#L162 and https://github.com/jenkinsci/ec2-plugin/blob/69ccf14155a7b06b1ad0675baaff2ca00c3f3f0b/src/main/java/hudson/plugins/ec2/util/EC2KeyPairManager.java#L19 are contradictory. And `Descriptor.save` is not saving this field; that is part of `Jenkins.clouds`.

At any rate, the new singleton will be inherently incompatible with CloudBees CI HA.

Reply:
i am not sure how the state seems to be persisted then (but i can clearly see that it is in the logs)